### PR TITLE
[Feature] New model for screening questions

### LINKED ISSUE #9038: ✨ Create new model for Screening Questions
## ✨ Feature

Create a new model for Screening Questions, with queries and mutations, and update the Assessment Plan builder UI to work with this model instead of the the renamed General Questions (#9034).

## 🕵️ Details

The distinction between screening questions and general questions is that screening questions are explicitly associated with skills, allowing them to be used for skill-based assessment. 

The UI should not need to be updated, despite the backend change. That should be taken care of in https://github.com/GCTC-NTGC/gc-digital-talent/issues/8161.

## 🎨 Data Model

This data model associates Screening Questions with Assessment Steps. (This technically allows for a pool to have multiple SCREEING_QUESTION-type steps, each associated with different sets of skills, which is not in the current design.)

We must be careful to only associate Screening Questions with Steps of the appropriate type. We can do validation in the mutation which creates Screening Questions.

For ease-of-use on the API, we should add a screeningQuestions field to the Pool which handles the hasManyThrough relationship implicitly.

```mermaid
classDiagram

class Pool { }

class PoolSkill { }

class Skill { }

class AssessmentStep {
    Int sortOrder
    AssessmentType type
    LocalizedString title
}

class GeneralQuestion {
    Int sortOrder
}

class GeneralQuestionResponse {
    String answer
}

class PoolCandidate

class ScreeningQuestion {
     Int sortOrder
}

class ScreeningQuestionResponse {
    String answer
}

Pool "1" --> "0..*" PoolSkill
PoolSkill "0..*" o-- "1" Skill

Pool "1" --o "0..*" AssessmentStep
AssessmentStep "0..*" -- "1..*" PoolSkill

Pool "1" --> "0..*" GeneralQuestion
GeneralQuestion "1" --> "1" GeneralQuestionResponse
PoolCandidate "1" --> "0..*" GeneralQuestionResponse
Pool "1" --o "0..*" PoolCandidate

AssessmentStep "1" --> "0..*" ScreeningQuestion : Only for SCREENING_QUESTION type steps
PoolCandidate "1" --> "0..*" ScreeningQuestionResponse
ScreeningQuestion "1" --> "1" ScreeningQuestionResponse
```

### 🧑‍🎨 Designer

Data model: @petertgiles / @tristan-orourke 

UI: @Jerryescandon 

## 📸 Screenshot

Here's the UI for where Screening Questions are added/edited, though it should already be complete.

https://www.figma.com/file/em5ArPr044YaIe1Vb1zCZz/Pools---Record-of-Decision?type=design&node-id=5737-16483&mode=design&t=NkgqlPqRQpQxhEOk-0

## 🙋‍♀️ Proposed Implementation

Migration for two new tables, with foreign keys.
PHP Models for new relationships.
Factories and seeders.
Add an extra relationship from Pools to Screening Questions (through Assessment Steps) and add it to the API schema as well.
Mutations to create or edit screening questions. (Validation of Assessment Step type.)
Querying or mutating screening questions check Assessment Plan permissions.

Notes about the API Schema

```
type AssessmentStep {
   ...
   screeningQuestions: [ScreeningQuestion] # if you have permission to see an assessment step, you have permission to see associated questions
}

type Pool {
   ...
   assessmentSteps: [AssessmentStep] # requires special permissions
   screeningQuestions: [ScreeningQuestion] # Applicants can see screening questions without going through assessment steps. Does NOT require special permissions, since anyone applying to the job has to see these.
}

type ScreeningQuestion {
   question: String
   assessmentStep: AssessmentStep # requires special permissions, because applicants can't see the full assessment plan
   responses: [ScreeningQuestionResponse] # DEFINITELY requires special permissions, because this is actually part of an application. Requires the same permissions as viewing PoolCandidate. We could leave this out of the API entirely, if we need to.
}
```

## 🌎 Localization

(optional) Provide any new copy along with translations available.

## ✅ Acceptance Criteria

A set of assumptions which, when tested, verify that the feature was properly implemented.

- [ ] New Screening Question and ScreeningQuestionResponse models exist, distinct from General Questions, matching the class diagram in this ticket
- [ ] Except for their relationships, the new models follow General Questions, with a sort order and localizable question.
- [ ] Assessment Plan builder manipulates Screening Questions, not General Questions
- [ ] phpunit tests for new mutations, particularly to test validation and permissions
- [ ] Its possible to query for Pool.screeningQuestions without going through assessment steps, and this is accessible to anyone looking at the Pool.
- [ ]  "completeness" validation logic in the frontend assessment plan builder and pool pages makes skills mandatory for every step

### PR DESCRIPTION
**Pull request by vd1992**
🤖 Resolves #9038 

## 👋 Introduction

Re-introduces screening questions.

## 🕵️ Details

Adds the screening question and screening question response models.
Screening questions re-uses the custom mutation class, screening responses copies general responses which itself was originally screening responses after all.
Screening question are associated with an assessment step as well as a pool.
Added a bit of PHP testing. 
Frontend assessment plan back to working with screening questions.
Screening question assessment step no longer has skills as optional. 

## 🧪 Testing

1. A fair bit going on
2. Migrate + seed successfully
3. Make a pool and step through the assessment plan builder setting screening questions

## 📸 Screenshot

image

### CODE REVIEW

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Looks great! Just a few small things.

File: api/database/migrations/2024_01_31_193707_add_new_screening_questions_table.php [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+    /**
+     * Reverse the migrations.
+     */
+    public function down(): void
+    {
+        Schema::dropIfExists('screening_question_responses');
+        Schema::dropIfExists('screening_questions');
+
+        // make migration reversible by recreating the incorrect foreign key names
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/database/migrations/2024_01_31_193707_add_new_screening_questions_table.php:68]: Yeah sure, it is a little odd to look at
https://github.com/GCTC-NTGC/gc-digital-talent/pull/9161/commits/1cfdb6b1905b087a8ad41c8394fd43466b815d58

File: api/app/GraphQL/Mutations/CreateOrUpdateScreeningQuestionAssessmentStep.php [hunk 2]
```diff
@@ -43,7 +58,8 @@ public function __invoke($_, array $args)
 
                     $questionToUpdate->save();
                 } else {
-                    $pool->generalQuestions()->Create([
+                    $pool->screeningQuestions()->Create([
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/GraphQL/Mutations/CreateOrUpdateScreeningQuestionAssessmentStep.php]: It appears to not matter either way, huh 🤔 
Switched it in https://github.com/GCTC-NTGC/gc-digital-talent/pull/9161/commits/0a0c05ff424003f683475780004cc399942c73d8

File: api/app/Models/ScreeningQuestion.php [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+    /**
+     * The attributes that can be filled using mass-assignment.
+     *
+     * @var array
+     */
+    protected $fillable = [
+        'question',
+        'sort_order',
+        'pool_id',
+        'assessment_step_id',
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/Models/ScreeningQuestion.php]: Sure, removed it in https://github.com/GCTC-NTGC/gc-digital-talent/pull/9161/commits/10361218e69e1fcfc0ea8c9f635eeacce9b3b4e0
It appears the mass assignment enables more convenient seeding, so it needed some tweaks to get it to work right.

File: api/app/Models/ScreeningQuestionResponse.php [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    protected $casts = [];
+
+    /**
+     * The attributes that can be filled using mass-assignment.
+     *
+     * @var array
+     */
+    protected $fillable = [
+        'pool_candidate_id',
+        'screening_question_id',
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/Models/ScreeningQuestionResponse.php]: See above

File: api/graphql/schema.graphql [hunk 5]
```diff
@@ -162,6 +162,15 @@ type GeneralQuestion {
   # generalQuestionResponses: [GeneralQuestionResponse] @hasMany # TODO: uncomment this and add sensible permission check
 }
 
+type ScreeningQuestion {
+  id: ID!
+  assessmentStep: AssessmentStep
+    @belongsTo
+    @canOnParent(ability: "viewAssessmentStep")
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/graphql/schema.graphql]: Okay sure, I did it in https://github.com/GCTC-NTGC/gc-digital-talent/pull/9161/commits/ef97b250083dfb036335d464996c5469d7f240d6
I also noted that the view assessment collection in pool policy should probably mirror that, so also changed that in https://github.com/GCTC-NTGC/gc-digital-talent/pull/9161/commits/4bd4c85f18b20667fcdd2cf954af5369d10e051f

File: api/app/Policies/ScreeningQuestionPolicy.php [hunk 6]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace App\Policies;
+
+use App\Models\ScreeningQuestion;
+use App\Models\User;
+use Illuminate\Auth\Access\HandlesAuthorization;
+
+class ScreeningQuestionPolicy
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/app/Policies/ScreeningQuestionPolicy.php]: See below

File: api/graphql/schema.graphql (hunk 5 above)
[the remark this answers is not in the document]
Comment (tristan-orourke) [reply api/graphql/schema.graphql]: Good catch! While you're here, could you rename `PoolPolicy.viewAssessments` to `viewAssessmentPlan`? The name "viewAssessments" is ambiguous about whether it refers to Assessment Steps or Assessment Results.

File: api/app/Policies/AssessmentStepPolicy.php [hunk 7]
```diff
@@ -54,6 +54,22 @@ public function update(User $user, AssessmentStep $assessmentStep)
...
+     *
+     * @return \Illuminate\Auth\Access\Response|bool
+     */
+    public function view(User $user, AssessmentStep $assessmentStep)
+    {
+        if ($user->isAbleTo('view-any-assessmentPlan')) {
+            return true;
+        }
+
+        $assessmentStep->pool->loadMissing('team');
```
Comment (tristan-orourke) [inline api/app/Policies/AssessmentStepPolicy.php:68]:
```suggestion
        $assessmentStep->loadMissing('pool.team');
```

File: api/database/migrations/2024_01_31_193707_add_new_screening_questions_table.php [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+        // renaming reversed after foreign keys created
```
Comment (tristan-orourke) [inline api/database/migrations/2024_01_31_193707_add_new_screening_questions_table.php:69]:
```suggestion
        // Note: the renaming will be reversed after foreign keys are created
```

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Great job, especially on catching that change in Pool Policy. Just a couple minor requests now.

File: api/graphql/schema.graphql (hunk 5 above)
[the remark this answers is not in the document]
Comment (vd1992) [reply api/graphql/schema.graphql]: Yep, can do https://github.com/GCTC-NTGC/gc-digital-talent/pull/9161/commits/7dbd0e9bf1c8044655060c997a2c2ab1a4b28213

File: api/app/GraphQL/Mutations/CreateOrUpdateScreeningQuestionAssessmentStep.php [hunk 9]
```diff
@@ -43,7 +58,8 @@ public function __invoke($_, array $args)
 
                     $questionToUpdate->save();
                 } else {
-                    $pool->generalQuestions()->Create([
+                    $pool->screeningQuestions()->create([
+                        'assessment_step_id' => $assessmentStep->id,
```
Comment (tristan-orourke) [inline api/app/GraphQL/Mutations/CreateOrUpdateScreeningQuestionAssessmentStep.php:62]: This doesn't work now with assessment_step_id taken out of the `fillable` array. You'll need to create a ScreeningQuestions object, set assessment_step_id, then save it to the pool.

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
After changing the fillable array like I asked, the mutation doesn't breaks :(

File: api/app/GraphQL/Mutations/CreateOrUpdateScreeningQuestionAssessmentStep.php (hunk 9 above)
Comment (vd1992) [reply api/app/GraphQL/Mutations/CreateOrUpdateScreeningQuestionAssessmentStep.php]: Mass assignment sure is convenient. 
I had to make the fields nullable so the process would work. Then added a test afterwards for it. 
Assertion `assertSuccessful()` incorrectly returns true when the mutation fails off mass assignment which tripped me up for a bit, so I tested cautiously around that. 

https://github.com/GCTC-NTGC/gc-digital-talent/pull/9161/commits/6c4e2667f7b77bb1eead3b827e0e5e470632e9fc
https://github.com/GCTC-NTGC/gc-digital-talent/pull/9161/commits/69ec9749b4c9f0279534be1808a3204eee5a2914

File: api/app/GraphQL/Mutations/CreateOrUpdateScreeningQuestionAssessmentStep.php [hunk 10]
```diff
@@ -58,11 +58,12 @@ public function __invoke($_, array $args)
...
-                        'assessment_step_id' => $assessmentStep->id,
-                        'question' => $incomingQuestion['question'],
-                        'sort_order' => isset($incomingQuestion['sortOrder']) ? $incomingQuestion['sortOrder'] : null,
-                    ]);
+                    $newQuestion = new ScreeningQuestion;
+                    $newQuestion->assessment_step_id = $assessmentStep->id;
+                    $newQuestion->question = $incomingQuestion['question'];
+                    $newQuestion->sort_order = isset($incomingQuestion['sortOrder']) ? $incomingQuestion['sortOrder'] : null;
+                    $newQuestion->save();
+                    $pool->screeningQuestions()->save($newQuestion);
```
Comment (tristan-orourke) [inline api/app/GraphQL/Mutations/CreateOrUpdateScreeningQuestionAssessmentStep.php:66]: Here's how you can make this work without making pool_id nullable.
```suggestion
                    $newQuestion = new ScreeningQuestion();
                    $newQuestion->assessment_step_id = $assessmentStep->id;
                    $newQuestion->question = $incomingQuestion['question'];
                    $newQuestion->sort_order = isset($incomingQuestion['sortOrder']) ? $incomingQuestion['sortOrder'] : null;
                    $pool->screeningQuestions()->save($newQuestion);
```

**Review by tristan-orourke — CHANGES_REQUESTED** [review]
Thanks for adding the test! I don't actually want you to make the foreign ids nullable. I left a suggestion showing how to make it work without doing that.

File: api/app/GraphQL/Mutations/CreateOrUpdateScreeningQuestionAssessmentStep.php (hunk 10 above)
Comment (vd1992) [reply api/app/GraphQL/Mutations/CreateOrUpdateScreeningQuestionAssessmentStep.php]: Oh cool, thanks I didn't know you could do that. 
I implemented both things in https://github.com/GCTC-NTGC/gc-digital-talent/pull/9161/commits/727e790fdd913738b717466744d3f48c181c88b1

File: api/database/migrations/2024_01_31_193707_add_new_screening_questions_table.php [hunk 11]
```diff
@@ -30,9 +30,9 @@ public function up(): void
 
         Schema::create('screening_questions', function (Blueprint $table) {
             $table->uuid('id')->primary('id')->default(new Expression('gen_random_uuid()'));
-            $table->uuid('pool_id');
+            $table->uuid('pool_id')->nullable();
             $table->foreign('pool_id')->references('id')->on('pools');
-            $table->uuid('assessment_step_id');
+            $table->uuid('assessment_step_id')->nullable();
```
[the remark this answers is not in the document]
Comment (vd1992) [reply api/database/migrations/2024_01_31_193707_add_new_screening_questions_table.php]: https://github.com/GCTC-NTGC/gc-digital-talent/pull/9161/commits/727e790fdd913738b717466744d3f48c181c88b1

**Review by tristan-orourke — APPROVED** [review]
👍

**Pull request merged**